needy/build_cache.py

```diff
@@ -0,0 +1,108 @@
+import logging
+import os
+import time
+import json
+
+from contextlib import contextmanager
+
+from .cache import Cache
+from .filesystem import TempDir, dict_file
+from .manifest import Manifest
+
+
+class BuildCache:
+    def __init__(self, cache, lock_timeout=10, lifetime=2*7*24*60*60, gc_frequency=24*60*60):
+        self.__cache = cache
+        self.__lock_timeout = lock_timeout
+        self.__lifetime = lifetime
+        self.__gc_frequency = gc_frequency
+
+        with self.__load_cache_policies() as p:
+            if 'object_lifetime' not in p:
+                p['object_lifetime'] = self.__lifetime
+            if 'gc_frequency' not in p:
+                p['gc_frequency'] = self.__gc_frequency
+
+            self.__lifetime = p['object_lifetime']
+            self.__gc_frequency = p['gc_frequency']
+
+    @staticmethod
+    def manifest_key():
+        return '.manifest'
+
+    @staticmethod
+    def policy_key():
+        return '.policy'
+
+    @contextmanager
+    def __load_manifest(self):
+        try:
+            with self.__cache.lease_file(BuildCache.manifest_key(), timeout=self.__lock_timeout, create=True) as path:
+                with Manifest(path) as m:
+                    yield m
+        except:
+            logging.warn('cache unavailable: unable to load manifest')
+            raise
+
+    @contextmanager
+    def __load_cache_policies(self):
+        try:
+            with self.__cache.lease_file(BuildCache.policy_key(), timeout=self.__lock_timeout, create=True) as path:
+                with dict_file(path) as d:
+                    yield d
+        except:
+            logging.warn('cache unavailable: unable to load manifest')
+            raise
+
+    def manifest(self):
+        '''Returns manifest at call time. Manifest may change between calls'''
+        with self.__load_manifest() as m:
+            return m
+
+    def store_artifacts(self, directory, key):
+        '''Store artifacts to key. Makes liberal use of exceptions for errors.'''
+        logging.info('Storing build artifacts in cache')
+        try:
+            with self.__load_manifest() as m:
+                self.__cache.store_directory(directory, key, timeout=self.__lock_timeout)
```

Comment:
you could reduce contention if you update the manifest, then release the lock before the store

```
            with self.__load_manifest() as m:
                m.touch(key)
                self.__collect_garbage(m)
            self.__cache.store_directory(directory, key, timeout=self.__lock_timeout)
```